### DISCUSSION TITLE
将h264裸流封装为fmp4用于流式传输

### DISCUSSION THREAD

**Discussion by nnnpa31** [yapingcat/gomedia#50] [Q&A] [answered]
如#49 所说，我已经在您的指示下成功将h264的NALU包封装为mp4视频，也成功封装成为fmp4视频了。
我后来要做的，是把fmp4数据通过websocket流式地发送到网页，用于实时直播。
为了尽可能的低延迟，我要把每一帧都独立地封装为一个fmp4帧，以不考虑带宽的方式来实现需求。

对于gomedia来说，封装h264裸流并定义时间戳后，它会直接以阻塞的方式读写到单个文件内。由于h264裸流和向前端发送的fmp4流都是实时生成的，要达到目的，我应该如何封装fmp4，并流式发送呢？

**Comment by yapingcat** [comment, marked as answer]
> 如#49 所说，我已经在您的指示下成功将h264的NALU包封装为mp4视频，也成功封装成为fmp4视频了。 我后来要做的，是把fmp4数据通过websocket流式地发送到网页，用于实时直播。 为了尽可能的低延迟，我要把每一帧都独立地封装为一个fmp4帧，以不考虑带宽的方式来实现需求。
> 
> 对于gomedia来说，封装h264裸流并定义时间戳后，它会直接以阻塞的方式读写到单个文件内。由于h264裸流和向前端发送的fmp4流都是实时生成的，要达到目的，我应该如何封装fmp4，并流式发送呢？

1. mp4 muxer 可以自定义io，只要实现了 io.WriteSeeker 接口就可以，可以在你定义的io中实现流式发送
2. 对于单帧切片，可以每写入一帧之后，调用FlushFragment接口

**Reply by yapingcat** [reply]
可以参考 [example_mux_mp4_memory_io](https://github.com/yapingcat/gomedia/blob/main/example/example_mux_mp4_memory_io.go)
对于一个切片，读写都在内存中操作，调用FlushFragment之后通过ws发送给浏览器
